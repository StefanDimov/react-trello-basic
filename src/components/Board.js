import React from 'react'
import List from './List'
import { Navbar, Grid, Row, Col, Button } from 'react-bootstrap'

export default class Board extends React.Component {
    render() {
        const { title, lists, onCardClick, onCreateCardClick } = this.props

        const listElements = lists.map(list =>
            <Col key={list.title} sm={3}>
                <List title={list.title} cards={list.cards} onCardClick={onCardClick}>
                    <Button bsStyle="primary" block onClick={() => onCreateCardClick(list.id)}>Add Card</Button>
                </List>
            </Col>
        )

        return (
            <div>
                <Navbar fluid={true}>
                    <Navbar.Header>
                        <Navbar.Brand>{title}</Navbar.Brand>
                    </Navbar.Header>
                </Navbar>

                <Grid fluid={true}>
                    <Row>{listElements}</Row>
                </Grid>
            </div>
        )
    }
}

const { string, array, func } = React.PropTypes
Board.propTypes = {
    title: string.isRequired,
    lists: array.isRequired,
    onCardClick: func.isRequired,
    onCreateCardClick: func.isRequired
}
