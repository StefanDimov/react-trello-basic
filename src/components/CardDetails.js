import React from 'react'

export default class CardDetails extends React.Component {
    render() {
        return (
            <div>
                Card Details
            </div>
        )
    }
}
