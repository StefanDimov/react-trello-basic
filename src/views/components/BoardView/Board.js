import React from 'react'
import { Navbar, Nav, NavItem, Button, Glyphicon } from 'react-bootstrap'

import * as boardUtils from '../../../utils/boardUtils'
import CardList from './CardList'
import ListsWrapper from './ListsWrapper'
import AddNewListInput from './AddNewListInput'

/**
 * Component that visualizes a Board's lists and holds all the functionality regarding them.
 * @class comp/Board
 * @param {object} board The Board to be visualized
 * @param {function} onCardClick callback for when a card is clicked
 * @param {function} onCreateCard callback to which the data for a new card will be passed
 * @param {function} onCreateNewList callback to which the data for a new list will be passed
 * @param {function} onSaveBoard callback to which the data for updating a board will be passed
 */
export default class Board extends React.Component {
    render() { // eslint-disable-line
        const { board, onCardClick, onCreateCard, onCreateNewList, onSaveBoard } = this.props
        const { title, lists } = board

        const listElements = lists.map(list =>
            <CardList key={list.id} list={list} onCardClick={onCardClick}>
                <Button bsStyle="primary" block onClick={() => onCreateCard(list.id)}>Add Card</Button>
            </CardList>
        )

        return (
            <div className="my-navbar-top-padding">
                <Navbar fluid={true} fixedTop={true}>
                    <Navbar.Header>
                        <Navbar.Brand>{title}</Navbar.Brand>
                    </Navbar.Header>
                    <Nav pullRight={true}>
                        <NavItem onClick={() => { onSaveBoard(board) }}>
                            <Glyphicon glyph="floppy-disk" /> Save Board
                        </NavItem>
                        <NavItem href="/#">
                            <Glyphicon glyph="th-large" /> Boards
                        </NavItem>
                    </Nav>
                </Navbar>

                <ListsWrapper>
                    {listElements}
                    <AddNewListInput onCreateList={onCreateNewList} />
                </ListsWrapper>
            </div>
        )
    }
}

const { func } = React.PropTypes
Board.propTypes = {
    board: boardUtils.boardPropType.isRequired,
    onCardClick: func.isRequired,
    onCreateCard: func.isRequired,
    onCreateNewList: func.isRequired,
    onSaveBoard: func.isRequired
}
