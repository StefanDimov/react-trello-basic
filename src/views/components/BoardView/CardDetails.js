import React from 'react'
import R from 'ramda'
import * as cardUtils from '../../../utils/cardUtils'
import { Row, Col, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap'

/**
 * Component that visualizes a Card's data in inputs. Card's data can be edited.
 * There is a side menu with actions that can be performed on the card as well.
 * @class comp/CardDetails
 * @param {object} card The card data to be visualized
 * @param {function} onCardSave callback which is called with the data when a card needs to be saved
 * @param {function} onCardDelete callback which is called with the data when a card needs to be deleted
 * @param {function} onCardCopy callback which is called with the data when a card needs to be copied
 */
export default class CardDetails extends React.Component {
    constructor(props) { // eslint-disable-line
        super(props)

        this.state = R.clone(props.card)

        this.handleTitleChange = this.handleTitleChange.bind(this)
        this.handleDescriptionChange = this.handleDescriptionChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    /**
     * Handler for the change event of the title input
     * @private
     * @param {object} event event data passed to the handler
     */
    handleTitleChange(event) {
        this.setState({title: event.target.value})
    }

    /**
     * Handler for the change event of the description input
     * @private
     * @param {object} event event data passed to the handler
     */
    handleDescriptionChange(event) {
        this.setState({description: event.target.value})
    }

    /**
     * Handler for the submit event of the form
     * @private
     * @param {object} event event data passed to the handler
     */
    handleSubmit(event) {
        event.preventDefault()
        // doesn't do anything
        // just to disable page reload on hitting enter
    }

    /**
     * When the component will unmount it will check if the card's data
     * has been changed in any way, if so it will call the onCardSave callback.
     * Also when new card is used (no fields assigned) it won't go in if no
     * field has been changed.
     * @private
     */
    componentWillUnmount() {
        // if the card has been changed
        if (!R.equals(this.state, this.props.card)) {
            // save new card
            this.props.onCardSave(R.clone(this.state))
        }
    }

    render() { // eslint-disable-line
        return (
            <Row>
                <Col sm={9}>
                    <form onSubmit={this.handleSubmit}>
                        <FormGroup>
                            <ControlLabel htmlFor="card-details-title-input">Title</ControlLabel>
                            <FormControl
                                id="card-details-title-input"
                                type="text"
                                value={this.state.title}
                                placeholder="Title..."
                                onChange={this.handleTitleChange} />
                        </FormGroup>

                        <FormGroup>
                            <ControlLabel htmlFor="card-details-description-input">Description</ControlLabel>
                            <FormControl
                                id="card-details-description-input"
                                componentClass="textarea"
                                value={this.state.description}
                                placeholder="Description..."
                                onChange={this.handleDescriptionChange} />
                        </FormGroup>
                    </form>
                </Col>

                <Col sm={3}>
                    <ControlLabel>Actions</ControlLabel>
                    <Button
                        id="card-copy-button"
                        block={true}
                        onClick={() => this.props.onCardCopy(R.clone(this.state))}
                        bsStyle="primary" bsSize="sm">Copy</Button>

                    <Button
                        id="card-delete-button"
                        block={true}
                        onClick={() => this.props.onCardDelete(R.clone(this.state))}
                        bsStyle="danger" bsSize="sm">Delete</Button>
                </Col>
            </Row>
        )
    }
}

CardDetails.propTypes = {
    card: cardUtils.cardPropType.isRequired,
    onCardSave: React.PropTypes.func.isRequired,
    onCardDelete: React.PropTypes.func.isRequired,
    onCardCopy: React.PropTypes.func.isRequired
}
