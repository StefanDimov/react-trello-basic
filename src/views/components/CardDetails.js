import React from 'react'
import R from 'ramda'
import cardUtils from '../../utils/cardUtils'
import { Grid, Row, Col, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap'

export default class CardDetails extends React.Component {
    constructor(props) {
        super(props)

        this.state = R.clone(props.card)

        this.handleTitleChange = this.handleTitleChange.bind(this)
        this.handleDescriptionChange = this.handleDescriptionChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    handleTitleChange(event) {
        this.setState({title: event.target.value})
    }

    handleDescriptionChange(event) {
        this.setState({description: event.target.value})
    }

    handleSubmit(event) {
        event.preventDefault();
        // doesn't do anything
        // just to disable page reload on hitting enter
    }

    componentWillUnmount() {
        // if the card has been changed
        // also when new card: it won't go in if no field has been changed
        if (!R.equals(this.state, this.props.card)) {
            // save new card
            this.props.onCardSave(R.clone(this.state));
        }
    }

    render() {
        return (
            <Row>
                <Col sm={9}>
                    <form onSubmit={this.handleSubmit}>
                        <FormGroup>
                            <ControlLabel htmlFor="card-details-title-input">Title</ControlLabel>
                            <FormControl
                                id="card-details-title-input"
                                type="text"
                                value={this.state.title}
                                placeholder="Title..."
                                onChange={this.handleTitleChange} />
                        </FormGroup>

                        <FormGroup>
                            <ControlLabel htmlFor="card-details-description-input">Description</ControlLabel>
                            <FormControl
                                id="card-details-description-input"
                                componentClass="textarea"
                                value={this.state.description}
                                placeholder="Description..."
                                onChange={this.handleDescriptionChange} />
                        </FormGroup>
                    </form>
                </Col>

                <Col sm={3}>
                    <ControlLabel>Actions</ControlLabel>
                    <Button
                        id="card-copy-button"
                        block={true}
                        onClick={() => this.props.onCardCopy(R.clone(this.state))}
                        bsStyle="primary" bsSize="sm">Copy</Button>

                    <Button
                        id="card-delete-button"
                        block={true}
                        onClick={() => this.props.onCardDelete(R.clone(this.state))}
                        bsStyle="danger" bsSize="sm">Delete</Button>
                </Col>
            </Row>
        )
    }
}

CardDetails.propTypes = {
    card: cardUtils.cardPropType.isRequired,
    onCardSave: React.PropTypes.func.isRequired,
    onCardDelete: React.PropTypes.func.isRequired,
    onCardCopy: React.PropTypes.func.isRequired
}
