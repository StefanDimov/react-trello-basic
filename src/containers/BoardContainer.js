import React from 'react'
import Board from '../components/Board'
import cardUtils from '../utils/cardUtils'
import CardDetails from '../components/CardDetails'
import { Modal } from 'react-bootstrap'

export default class BoardContainer extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            showCardDetails: false,
            cardToView: null
        }

        this.viewCardDetails = this.viewCardDetails.bind(this)
        this.createCard = this.createCard.bind(this)
        this.saveCard = this.saveCard.bind(this)
        this.closeCardDetailsModal = this.closeCardDetailsModal.bind(this)
        this.createNewList = this.createNewList.bind(this)
    }

    viewCardDetails(card) {
        this.setState({
            showCardDetails: true,
            cardToView: card
        })
    }

    createCard(listId) {
        this.setState({
            showCardDetails: true,
            cardToView: cardUtils.createEmptyCard(listId)
        })
    }

    createNewList(listName) {
        console.log(listName);
    }

    closeCardDetailsModal() {
        this.setState({
            showCardDetails: false,
            cardToView: null
        })
    }

    saveCard(newCard) {
        // TODO: Implement: Save card action
        console.log(newCard);
    }

    render() {
        const { boardData } = this.props

        return (
            <div>
                <Board
                    title={boardData.title}
                    lists={boardData.lists}
                    onCardClick={this.viewCardDetails}
                    onCreateCardClick={this.createCard}
                    onCreateNewList={this.createNewList} />

                <Modal show={this.state.showCardDetails} onHide={this.closeCardDetailsModal} animation={false}>
                    <Modal.Body>
                        {this.state.cardToView &&
                            <CardDetails card={this.state.cardToView} onCardSave={this.saveCard} />}
                    </Modal.Body>
                </Modal>
            </div>
        )
    }
}
