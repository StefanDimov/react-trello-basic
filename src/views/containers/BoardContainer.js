import React from 'react'
import { Modal } from 'react-bootstrap'

import cardUtils from '../../utils/cardUtils'

import boardStore from '../../stores/boardStore'
import * as boardActions from '../../actions/boardActions'

import Board from '../components/Board'
import CardDetails from '../components/CardDetails'

export default class BoardContainer extends React.Component {
    constructor(props) {
        super(props)

        this.setBoard = this.setBoard.bind(this)
        this.viewCardDetails = this.viewCardDetails.bind(this)
        this.initCreateCard = this.initCreateCard.bind(this)
        this.saveCard = this.saveCard.bind(this)
        this.deleteCard = this.deleteCard.bind(this)
        this.closeCardDetailsModal = this.closeCardDetailsModal.bind(this)
        this.createNewList = this.createNewList.bind(this)

        this.state = {
            board: boardStore.getBoard(),
            showCardDetails: false,
            cardToView: null
        }
    }

    componentWillMount() {
        // subscribes to store
        boardStore.on('change', this.setBoard)
    }

    componentWillUnmount() {
        // unsubscribes to store
        boardStore.unbindListener('change', this.setBoard)
    }

    setBoard() {
        // sets board state from store
        this.setState({ board: boardStore.getBoard() })
    }

    // opens CardDetails with card
    viewCardDetails(card) {
        this.setState({
            showCardDetails: true,
            cardToView: card
        })
    }

    // closes CardDetails and clears state
    closeCardDetailsModal() {
        this.setState({
            showCardDetails: false,
            cardToView: null
        })
    }

    // opens CardDetails with new empty card
    initCreateCard(listId) {
        this.setState({
            showCardDetails: true,
            cardToView: cardUtils.createEmptyCard(listId)
        })
    }

    createNewList(listName) {
        boardActions.addNewList(listName)
    }

    saveCard(card) {
        boardActions.saveCard(card)
    }

    deleteCard(card) {
        boardActions.deleteCard(card)
        this.closeCardDetailsModal()
    }

    render() {
        const { board } = this.state

        return (
            <div>
                <Board
                    board={board}
                    onCardClick={this.viewCardDetails}
                    onCreateCardClick={this.initCreateCard}
                    onCreateNewList={this.createNewList} />

                <Modal show={this.state.showCardDetails} onHide={this.closeCardDetailsModal} animation={false}>
                    <Modal.Body>
                        {this.state.cardToView &&
                            <CardDetails card={this.state.cardToView} onCardSave={this.saveCard} onCardDelete={this.deleteCard} />}
                    </Modal.Body>
                </Modal>
            </div>
        )
    }
}
