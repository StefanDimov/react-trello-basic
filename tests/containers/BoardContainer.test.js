import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'


import BoardContainer from '../../src/containers/BoardContainer'
import Board from '../../src/components/Board'
import CardDetails from '../../src/components/CardDetails'
import { Modal } from 'react-bootstrap'

describe('BoardContainer', () => {

    let emptyBoard, boardWithTitle, boardWithTitleAndOneList, boardWithTitleAndTwoLists, card

    beforeEach(() => {
        emptyBoard = { id: 'uniqueId', title: '', lists: [] }
        boardWithTitle = { id: 'uniqueId', title: 'BoardTitle', lists: [] }
        boardWithTitleAndOneList = {
            id: 'uniqueId',
            title: 'BoardTitle',
            lists: [{ id: 'uniqueId', title: 'list', cards: [] }]
        }
        boardWithTitleAndTwoLists = {
            id: 'uniqueId',
            title: 'BoardTitle',
            lists: [
                { id: 'uniqueId', title: 'list', cards: [] },
                { id: 'uniqueId2', title: 'list2', cards: [] }
            ]
        }
        card = { id: 'uniqueId', title: 'cardTitle' }
    })

    describe('rendering', () => {

        it('should render correctly an empty board', () => {
            const tree = renderer.create(
                <BoardContainer board={emptyBoard}/>
            ).toJSON()

            expect(tree).toMatchSnapshot();
        })

        it('should render correctly a board with a title', () => {
            const tree = renderer.create(
                <BoardContainer board={boardWithTitle}/>
            ).toJSON()

            expect(tree).toMatchSnapshot();
        })

        it('should render correctly a board with a title and a list', () => {
            const tree = renderer.create(
                <BoardContainer board={boardWithTitleAndOneList}/>
            ).toJSON()

            expect(tree).toMatchSnapshot();
        })

        it('should render correctly a board with a title and two lists', () => {
            const tree = renderer.create(
                <BoardContainer board={boardWithTitleAndTwoLists}/>
            ).toJSON()

            expect(tree).toMatchSnapshot();
        })
    })

    describe('card creation', () => {

        it('should open card creation modal with appropriate card and save function', () => {
            const wrapper = shallow(<BoardContainer board={emptyBoard} />)
            const instance = wrapper.instance()

            instance.viewCardDetails(card)

            expect(wrapper.find(Modal).prop('show')).toBe(true)
            expect(wrapper.find(Modal).find(CardDetails).length).toBe(1)
            expect(wrapper.find(Modal).find(CardDetails).prop('card')).toBe(card)
            expect(wrapper.find(Modal).find(CardDetails).prop('onCardSave')).toBe(instance.saveCard)
        })

        it('should hide card creation modal and clear the state', () => {
            const wrapper = shallow(<BoardContainer board={emptyBoard} />)
            const instance = wrapper.instance()

            instance.viewCardDetails(card)
            expect(wrapper.find(Modal).prop('show')).toBe(true)
            expect(wrapper.find(Modal).find(CardDetails).length).toBe(1)

            expect(wrapper.find(Modal).prop('onHide')).toBe(instance.closeCardDetailsModal)
            instance.closeCardDetailsModal()
            expect(wrapper.find(Modal).prop('show')).toBe(false)
            expect(wrapper.find(Modal).find(CardDetails).length).toBe(0)
        })

        it('should call card creation modal with empty card', () => {
            const wrapper = shallow(<BoardContainer board={emptyBoard} />)
            const instance = wrapper.instance()
            const listId = 'someId'

            instance.initCreateCard(listId)
            expect(wrapper.find(Modal).prop('show')).toBe(true)
            expect(wrapper.find(Modal).find(CardDetails).length).toBe(1)

            expect(wrapper.find(Modal).prop('onHide')).toBe(instance.closeCardDetailsModal)
            expect(wrapper.find(Modal).find(CardDetails).prop('onCardSave')).toBe(instance.saveCard)
            expect(wrapper.find(Modal).find(CardDetails).prop('card').listId).toBe(listId)
            expect(wrapper.find(Modal).find(CardDetails).prop('card').title).toBe('')
            expect(wrapper.find(Modal).find(CardDetails).prop('card').description).toBe('')
        })
    })

    describe('Board', () => {
        it('should be passed the right functions', () => {
            const wrapper = shallow(<BoardContainer board={emptyBoard} />)
            const instance = wrapper.instance()

            expect(wrapper.find(Board).prop('onCardClick')).toBe(instance.viewCardDetails)
            expect(wrapper.find(Board).prop('onCreateCardClick')).toBe(instance.initCreateCard)
            expect(wrapper.find(Board).prop('onCreateNewList')).toBe(instance.createNewList)
        })
    })
})
