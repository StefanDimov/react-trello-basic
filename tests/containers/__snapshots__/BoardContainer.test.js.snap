exports[`BoardContainer rendering should render correctly a board with a title 1`] = `
<div>
  <div
    className="my-navbar-top-padding">
    <nav
      className="navbar navbar-default navbar-fixed-top">
      <div
        className="container-fluid">
        <div
          className="navbar-header">
          <span
            className="navbar-brand">
            BoardTitle
          </span>
        </div>
      </div>
    </nav>
    <div
      className="my-lists-container">
      <form
        className="my-list-container-item"
        onSubmit={[Function]}>
        <div
          className="form-group">
          <input
            className="form-control"
            id={undefined}
            onChange={[Function]}
            placeholder="Add new list..."
            type="text"
            value="" />
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`BoardContainer rendering should render correctly a board with a title and a list 1`] = `
<div>
  <div
    className="my-navbar-top-padding">
    <nav
      className="navbar navbar-default navbar-fixed-top">
      <div
        className="container-fluid">
        <div
          className="navbar-header">
          <span
            className="navbar-brand">
            BoardTitle
          </span>
        </div>
      </div>
    </nav>
    <div
      className="my-lists-container">
      <div
        className="my-list-container-item panel panel-default"
        id={undefined}>
        <div
          className="panel-heading">
          list
        </div>
        <div
          className="panel-body">
          <button
            className="btn btn-primary btn-block"
            disabled={false}
            onClick={[Function]}
            type="button">
            Add Card
          </button>
        </div>
      </div>
      <form
        className="my-list-container-item"
        onSubmit={[Function]}>
        <div
          className="form-group">
          <input
            className="form-control"
            id={undefined}
            onChange={[Function]}
            placeholder="Add new list..."
            type="text"
            value="" />
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`BoardContainer rendering should render correctly a board with a title and two lists 1`] = `
<div>
  <div
    className="my-navbar-top-padding">
    <nav
      className="navbar navbar-default navbar-fixed-top">
      <div
        className="container-fluid">
        <div
          className="navbar-header">
          <span
            className="navbar-brand">
            BoardTitle
          </span>
        </div>
      </div>
    </nav>
    <div
      className="my-lists-container">
      <div
        className="my-list-container-item panel panel-default"
        id={undefined}>
        <div
          className="panel-heading">
          list
        </div>
        <div
          className="panel-body">
          <button
            className="btn btn-primary btn-block"
            disabled={false}
            onClick={[Function]}
            type="button">
            Add Card
          </button>
        </div>
      </div>
      <div
        className="my-list-container-item panel panel-default"
        id={undefined}>
        <div
          className="panel-heading">
          list2
        </div>
        <div
          className="panel-body">
          <button
            className="btn btn-primary btn-block"
            disabled={false}
            onClick={[Function]}
            type="button">
            Add Card
          </button>
        </div>
      </div>
      <form
        className="my-list-container-item"
        onSubmit={[Function]}>
        <div
          className="form-group">
          <input
            className="form-control"
            id={undefined}
            onChange={[Function]}
            placeholder="Add new list..."
            type="text"
            value="" />
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`BoardContainer rendering should render correctly an empty board 1`] = `
<div>
  <div
    className="my-navbar-top-padding">
    <nav
      className="navbar navbar-default navbar-fixed-top">
      <div
        className="container-fluid">
        <div
          className="navbar-header">
          <span
            className="navbar-brand">
            
          </span>
        </div>
      </div>
    </nav>
    <div
      className="my-lists-container">
      <form
        className="my-list-container-item"
        onSubmit={[Function]}>
        <div
          className="form-group">
          <input
            className="form-control"
            id={undefined}
            onChange={[Function]}
            placeholder="Add new list..."
            type="text"
            value="" />
        </div>
      </form>
    </div>
  </div>
</div>
`;
